astropy/io/ascii/core.py

```diff
@@ -31,14 +34,17 @@
 from . import connect
 from .docs import READ_DOCSTRING, WRITE_DOCSTRING
 
+if TYPE_CHECKING:
+    from typing import ClassVar, Final, Literal, Self, SupportsFloat
+
 # Global dictionary mapping format arg to the corresponding Reader class
-FORMAT_CLASSES = {}
+FORMAT_CLASSES: dict[str, MetaBaseReader] = {}
```

Comment:
I may be confused by the presence of a metaclass here, but I would expect this to be 
```suggestion
FORMAT_CLASSES: dict[str, BaseReader] = {}
```

Reply:
The machinery that updates these dictionaries is in `MetaBaseReader`. The type system cannot assume that only `BaseReader` inherits from `MetaBaseReader` directly.

Reply:
I'm still confused. Strictly speaking if a member of `FORMAT_CLASSES` was only `MetaBaseReader` instance by not `BaseReader` that would be incorrect.

Reply:
The dictionaries in question are updated in https://github.com/astropy/astropy/blob/76cbb5ad5dc4af1fefec84d9900e8ca03898326d/astropy/io/ascii/core.py#L1213-L1225
That code has no way of knowing that it is only called by `BaseReader` or its subclasses, so a type checker __must__ report using a narrower type for the values of the dictionaries as a type error.

I seems to me that developer intent would be better expressed by moving (at least) the code that updates the dictionaries from the metaclass to `BaseReader.__init_subclass__()`, in which case the dictionaries could indeed be annotated as `dict[str, type[BaseReader]]`.

You can look up how a similar dictionary of unit formatters is maintained by the `Base` formatter: https://github.com/astropy/astropy/blob/76cbb5ad5dc4af1fefec84d9900e8ca03898326d/astropy/units/format/base.py#L37-L44

Reply:
thanks for explaining ! I agree that refactoring using `__init_subclass__` would be clearer. Though it should probably be done as a separate PR (which I'd immediately approve)

Reply:
Turns out that `FastBasic` inherits directly from the metaclass: https://github.com/astropy/astropy/blob/19f5aa99321ddb42e5b01dc79159321762e0b2c5/astropy/io/ascii/fastbasic.py#L12
That means that moving the code in the metaclass to `BaseReader.__init_subclass__()` is not trivial, and I'd rather not get bogged down with rewriting `FastBasic` as a subclass of `BasicReader`. What it means for the code being discussed is that `dict[str, type[BasicReader]]` is too narrow and the annotation I have here is the correct one.

Reply:
alright, thanks for trying !